Add support for local overrides from .ddev.toml file

AAraKKe opened the pull request:
### What does this PR do?
_Note: PR opened to start a discussion and validate actions_

This PR modifies the `ConfiFile` class to support the main global configuration file but also check whether the current directory has a local `.ddev.toml` configuration file. If it is exist the model that represents the configuration file will be a merge of both in which anything in the local `.ddev.toml` file overrides the central one.

The config cli tools have been updated to addapt to this change:

#### Show
When a `.ddev.toml` file is present, running the `show` commnad will not only show the configuration but also mention the source of the configuration.

For a central config file like
<details>
<summary> Global config file </summary>

```toml
repo = "core"
agent = "dev"
org = "default"

[repos]
core = "~/dd/integrations-core"
extras = "~/dd/integrations-extras"
marketplace = "~/dd/marketplace"
agent = "~/dd/datadog-agent"

[agents.dev]
docker = "datadog/agent-dev:master"
local = "latest"

[agents.7]
docker = "datadog/agent:7"
local = "7"

[orgs.default]
api_key = ""
app_key = ""
site = "datadoghq.com"
dd_url = "https://app.datadoghq.com"
log_url = ""

[github]
user = ""
token = ""

[pypi]
user = ""
auth = ""

[trello]
key = ""
token = ""

[terminal.styles]
info = "bold"
success = "bold cyan"
error = "bold red"
warning = "bold yellow"
waiting = "bold magenta"
debug = "bold"
spinner = "simpleDotsScrolling"
```

</details>

and a local `.ddev.toml` like this one

<details>
<summary> Local `.ddev.toml` </summary>

```toml
repo = "local"

[repos]
local = "some path"
```

</details>

Running `ddev config show` outputs

<details>
<summary> Show command output </summary>

```toml
repo = "local"                         # .ddev.toml:0
agent = "dev"                          # config.toml:1
org = "default"                        # config.toml:2

[repos]                                # .ddev.toml:2
core = "~/dd/integrations-core"        # config.toml:5
extras = "~/dd/integrations-extras"    # config.toml:6
marketplace = "~/dd/marketplace"       # config.toml:7
agent = "~/dd/datadog-agent"           # config.toml:8
local = "some path"                    # .ddev.toml:3

[agents.dev]                           # config.toml:10
docker = "datadog/agent-dev:master"    # config.toml:11
local = "latest"                       # config.toml:12

[agents.7]                             # config.toml:14
docker = "datadog/agent:7"             # config.toml:15
local = "7"                            # config.toml:16

[orgs.default]                         # config.toml:18
api_key = "*****"                      # config.toml:19
app_key = "*****"                      # config.toml:20
site = "datadoghq.com"                 # config.toml:21
dd_url = "https://app.datadoghq.com"   # config.toml:22
log_url = ""                           # config.toml:23

[github]                               # config.toml:25
user = ""                              # config.toml:30
token = "*****"                        # config.toml:35

[pypi]                                 # config.toml:29
user = ""                              # config.toml:30
auth = "*****"                         # config.toml:31

[trello]                               # config.toml:33
key = ""                               # config.toml:34
token = "*****"                        # config.toml:35

[terminal.styles]                      # config.toml:37
info = "bold"                          # config.toml:38
success = "bold cyan"                  # config.toml:39
error = "bold red"                     # config.toml:40
warning = "bold yellow"                # config.toml:41
waiting = "bold magenta"               # config.toml:42
debug = "bold"                         # config.toml:43
spinner = "simpleDotsScrolling"        # config.toml:44
```

</details>

#### Edit & set
The `edit` and `set` command now include a `--local` option that allows to modify the local `.ddev.toml` file.

#### Restore
When running `restore` if the local file is present the user will be prompted whether the local file should be removed as well. This is expected if the user wants to restore the configuration to avoid the local file overriding any value. The user is prompted instead of deleting it because the use case of this file (override variables for specific projects) means that they might not want to remove it to recreate it later.

### Motivation
It has been raised a couple of times that moving from one project to another normally means that different copies of the repo should be opened at the same time (worktrees, different folders...). When this happen using ddev has some drawbacks given how the configuration works. The repo variable mostly in which we would need to change it or rely on side effects (like adding `$PWD` as the repo path which is expanded).

This is an attempt to simplify this.

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Add the `qa/skip-qa` label if the PR doesn't need to be tested during QA.
- [ ] If you need to backport this PR to another branch, you can add the `backport/<branch-name>` label to the PR and it will automatically open a backport PR once this one is merged

nubtron reviewed (approved):
Looks good to me! You might want to add `.ddev.toml` to the repo's `.gitignore`.

Small feature idea (non-blocking!): recurse into parent directories until you find a config, so that ddev can easily used with configs in subdirectories in the repo.

AAraKKe commented:
This makes a lot of sense @nubtron , I am working on adding support to traverse through parent directories to ensure we can run it from a subdirectory. Since it is very common to be the directory of a given integration.

Will add it to gitignore as well.

AAraKKe commented:
Just updated the logic to support traversing through parent directories and updated the documentation to include a new page with the workflow to work with multiple repos and/or worktrees.

nubtron commented on ddev/src/ddev/config/file.py:
```diff
@@ -3,71 +3,316 @@
-from ddev.utils.toml import load_toml_data
+from ddev.utils.toml import dumps_toml_data, load_toml_data
 
+LOCAL_OVERRIDES_PATH = Path.cwd() / ".ddev.toml"
```
I think this is no longer used.

nubtron commented on ddev/src/ddev/config/file.py:
```diff
@@ -3,71 +3,316 @@
-    def path(self, value):
-        self._path = value
+        # Keep searching while the file doesn't exist in the current directory
+        while not (current_dir / ".ddev.toml").is_file():
```
We should probably add a try/catch here in case we don't have permissions for a parent directory! And maybe a list of walked paths to avoid getting stuck in symlink cycles?

AAraKKe replied:
I added support for this case and covered it with unit tests. I also removed the `".ddev.toml"` from everywhere. Now is a constant. I wanted to do that before but forgot.

nubtron reviewed (approved):
Approved with a couple of notes about the loop that finds the config override!

nubtron commented on .cursor/rules/python-type-hinting.mdc:
```diff
@@ -0,0 +1,29 @@
+
+This way, whether a and b are either strings or bytes, they cannot be mixed. 
+
+If a single argument is present in the function, `str | bytes` is preferred.
```
Nice cursor files!

nubtron commented on ddev/src/ddev/config/model.py:
```diff
@@ -79,7 +79,7 @@ def repo(self):
                 self.raise_error('must be a string')
             elif repo not in self.repos:
-                self.raise_error('unknown repository')
+                self.raise_error(f'unknown Repository: {repo!r}')
```
This will be useful!

AAraKKe commented on ddev/src/ddev/cli/config/local_repo.py:
```diff
@@ -0,0 +1,104 @@
+
+@click.command()
+@click.pass_obj
+def local_repo(app: Application):
```
Change the name to this command to `override` (after call with Ilia)

michaelcretzman reviewed (approved):
Approved with the assumption that you will consider my edits. One edit corrects a spelling mistake. Please incorp that one.
